IdentityCore/src/webview/response/MSIDWebOAuth2Response.m

```diff
@@ -26,8 +26,127 @@
 //------------------------------------------------------------------------------
 
 #import "MSIDWebOAuth2Response.h"
+#import "MSIDWebAADAuthResponse.h"
+#import "MSIDWebWPJAuthResponse.h"
 
 @implementation MSIDWebOAuth2Response
 
++ (MSIDWebOAuth2Response *)responseWithURL:(NSURL *)url
+                              requestState:(NSString *)requestState
+                             stateVerifier:(MSIDWebUIStateVerifier)stateVerifier
+                                   context:(id<MSIDRequestContext>)context
+                                     error:(NSError **)error;
+
+{
+    // This error case *really* shouldn't occur. If we're seeing it it's almost certainly a developer bug
+    if ([NSString msidIsStringNilOrBlank:url.absoluteString])
+    {
+        if (error){
+            *error = MSIDCreateError(MSIDOAuthErrorDomain, MSIDErrorNoAuthorizationResponse, @"No authorization response received from server.", nil, nil, nil, context.correlationId, nil);
+            return nil;
+        }
+    }
+    
+    // Check for WPJ response
+    if([url.absoluteString hasPrefix:@"msauth://"])
+    {
+        NSString *query = [url query];
+        NSDictionary *queryParams = [NSDictionary msidURLFormDecode:query];
+        NSString *appURLString = [queryParams objectForKey:@"app_link"];
+        
+        MSIDWebWPJAuthResponse *response = [MSIDWebWPJAuthResponse new];
+        response.url = url;
+        response.appInstallLink = appURLString;
+        
+        return response;
+    }
+    
+    // Check for auth response
+    // Try both the URL and the fragment parameters:
+    NSDictionary *parameters = [url msidFragmentParameters];
+    if (parameters.count == 0)
+    {
+        parameters = [url msidQueryParameters];
+    }
+    
+    // Verify state
+    BOOL stateVerified = stateVerifier(parameters, requestState);
+    
+    if (!stateVerified)
+    {
+        if (error){
+            *error = MSIDCreateError(MSIDOAuthErrorDomain, MSALErrorInvalidState, @"State returned from the server does not match", nil, nil, nil, context.correlationId, nil);
+            return nil;
+        }
+    }
+    
+    NSString *code = parameters[MSID_OAUTH2_CODE];
+    NSString *cloudHostName = parameters[MSID_AUTH_CLOUD_INSTANCE_HOST_NAME];
```

Comment:
cloud host name is something AAD specific, let's discuss if it should be here

Reply:
Like above :)